system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -372,6 +393,396 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+                {
+                    if (required)
+                        throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: xpath node not found '%s' in xpath '%s'", finger, xpath);
+                    return false;
+                }
+                node = xmlNewChild(node, nullptr, (const xmlChar *) finger, nullptr);
+                if (!node)
+                {
+                    if (required)
+                        throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: error creating node '%s' in xpath '%s'", finger, xpath);
+                    return false;
+                }
+            }
+            else
+            {
+                xmlNodePtr *nodes = obj->nodesetval->nodeTab;
+                if (obj->nodesetval->nodeNr>1)
+                {
+                    if (required)
+                        throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: ambiguous xpath '%s' at node '%s'", xpath, finger);
+                    return false;
+                }
+                node = *nodes;
+            }
+            xmlXPathFreeObject(obj);
+        }
+        setLocation(node);
+        return true;
+    }
+
+    virtual void setLocationNamespace(const char *prefix, const char * uri, bool current) override
+    {
+        xmlNodePtr node = m_xpathContext->node;
+        if (!node)
+            return;
+        xmlNsPtr ns = nullptr;
+        if (isEmptyString(uri))
+            ns = xmlSearchNs(node->doc, node, (const xmlChar *) prefix);
+        else
+            ns = xmlNewNs(node, (const xmlChar *) uri, (const xmlChar *) prefix);
+        if (current && ns)
+            xmlSetNs(node, ns);
+    }
+
+    enum class ensureValueMode { set, add, appendto };
+    virtual void ensureValue(const char *xpath, const char *value, ensureValueMode action, bool required)
+    {
+        if (isEmptyString(xpath))
+            return;
+        if (strstr(xpath, "//")) //maybe in the future, scripting error rather than not found, required==false has no effect
+            throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureValue: '//' not currently allowed for set operations '%s'", xpath);
+
+        xmlNodePtr current = m_xpathContext->node;
+        if (!current || *xpath=='/')
+        {
+            current = xmlDocGetRootElement(m_xmlDoc); //valuable, but very dangerous
+            if (*xpath=='/')
+                xpath++;
+        }
+        xmlNodePtr node = current;
+
+        StringArray xpathnodes;
+        xpathnodes.appendList(xpath, "/");
+        ForEachItemIn(i, xpathnodes)
+        {
+            const char *finger = xpathnodes.item(i);
+            if (isEmptyString(finger)) //shouldn't happen, scripting error rather than not found, required==false has no effect
+                throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureValue: '//' not allowed '%s'", xpath);
```

Comment:
Possibly misleading message. You've already rejected any occurrence of "//", so this is caused by trimmed whitespace when splitting the path, or (possibly) a leading or trailing '/'. It would be better to distinguish this exception message from the one above.